cptv/writer.py

```diff
@@ -0,0 +1,231 @@
+        fw.uint32(ord(Field.X_RESOLUTION), COLS)
+        fw.uint32(ord(Field.Y_RESOLUTION), ROWS)
+
+        if self.device_name:
+            fw.string(ord(Field.DEVICENAME), self.device_name)
+
+        if self.device_id:
+            fw.uint32(ord(Field.DEVICEID), self.device_id)
+
+        fw.timestamp(ord(Field.TIMESTAMP), self.timestamp)
+
+        if self.latitude:
+            fw.float32(ord(Field.LATITUDE), self.latitude)
+
+        if self.longitude:
+            fw.float32(ord(Field.LONGITUDE), self.longitude)
+
+        if self.preview_secs:
+            fw.uint8(ord(Field.PREVIEW_SECS), self.preview_secs)
+
+        if self.motion_config:
+            fw.string(ord(Field.MOTION_CONFIG), self.motion_config)
+
+        fw.write(ord(Section.HEADER), self.s)
+
+    def write_frame(self, frame):
+        bit_width, start_value, frame_buf = self.comp._next_frame(frame.pix)
+
+        fw = FieldWriter()
+        fw.uint32(ord(Field.TIME_ON),
+                  frame.time_on / timedelta(milliseconds=1))
+        fw.uint32(ord(Field.LAST_FFC_TIME),
+                  frame.last_ffc_time / timedelta(milliseconds=1))
+        fw.uint8(ord(Field.BIT_WIDTH), bit_width)
+        fw.uint32(ord(Field.FRAME_SIZE), len(frame_buf) + 4)
+        fw.write(ord(Section.FRAME), self.s)
+
+        self.s.write(struct.pack("<l", start_value))
+        self.s.write(frame_buf)
+
+    def close(self):
+        self.s.close()
+
+
+class FieldWriter:
+
+    def __init__(self):
+        self.s = BytesIO()
+        self.count = 0
+
+    def write(self, section_type, dest):
+        dest.write(struct.pack("<BB", section_type, self.count))
+        dest.write(self.s.getbuffer())
+
+    def timestamp(self, code, t):
+        micros = int(t.timestamp() * 1e6)
+        self.uint64(code, micros)
+
+    def uint8(self, code, val):
+        self.s.write(struct.pack("<BBB", 1, code, val))
+        self.count += 1
+
+    def uint32(self, code, val):
+        self.s.write(struct.pack("<BBL", 4, code, int(val)))
+        self.count += 1
+
+    def uint64(self, code, val):
+        self.s.write(struct.pack("<BBQ", 8, code, val))
+        self.count += 1
+
+    def float32(self, code, fval):
+        self.s.write(struct.pack("<BBf", 4, code, fval))
+        self.count += 1
+
+    def string(self, code, val):
+        self.s.write(struct.pack("<BB", len(val), code))
+        self.s.write(val)
+        self.count += 1
+
+
+class Compressor:
+
+    def _get_twisted(self):
+        if not hasattr(self, 'twisted'):
+            width = COLS
+            height = ROWS
+            linear = np.arange(width * height, dtype='I')
+            self.twisted = linear + \
+                ((linear // width) & 1) * (width - 1 - 2 * (linear % width))
+        return self.twisted
+
+    def _next_frame(self, pix):
+        linear_pix = np.empty(COLS * ROWS, dtype='h')
+
+        twisted = self._get_twisted()
+        linear_pix = pix.ravel().astype('h')[twisted]
```

Comment:
Is `linear_pix` here not being reassigned before it is used?

Reply:
Yeah, that looks wrong.